File: Java/2023/Solvers/Day02Solver.java
package Solvers;

import java.io.IOException;
import java.util.ArrayList;

import Inputs.InputParser;

public class Day02Solver extends InputParser{
    public int day02Part01() throws IOException {
        ArrayList<String> games = getGames();
        Helpers helper = new Helpers();

        int answer = 0;

        for (String game : games) {
            int index = game.indexOf(":");
            String gameRolls = game.substring(index + 2);
            boolean isBadGame = false;

            StringBuilder sb = new StringBuilder();
            char[] charArray = gameRolls.toCharArray();
            int cubes = 0;

            for (char letter : charArray) {
                if (Character.isDigit(letter)) {
                    sb.append(letter);
                    cubes = Integer.parseInt(sb.toString());
                } else {
                    sb.setLength(0);
                }

                if ((cubes > 14 && letter == 'b') || (cubes > 13 && letter == 'n') || (cubes > 12 && letter == 'd')) {
                    isBadGame = true;
                    cubes = 0;
                    break;
                }
            }

            if (!isBadGame) {
                answer += helper.getLineId(game);
            }
        }

        return answer;
    }

    public int day02Part02() throws IOException {
        ArrayList<String> games = getGames();

        int answer = 0;

        for (String game : games) {
            int index = game.indexOf(":");
            String gameRolls = game.substring(index + 2);

            StringBuilder sb = new StringBuilder();
            char[] charArray = gameRolls.toCharArray();
            int cubes = 0;
            int blue = 0;
            int green = 0;
            int red = 0;

            for (char letter : charArray) {
                if (Character.isDigit(letter)) {
                    sb.append(letter);
                    cubes = Integer.parseInt(sb.toString());
                } else {
                    sb.setLength(0);
                }

                if (letter == 'b' && cubes > blue) {
                    blue = cubes;
                    cubes = 0;
                }

                if (letter == 'n' && cubes > green) {
                    green = cubes;
                    cubes = 0;
                }

                if (letter == 'd' && cubes > red) {
                    red = cubes;
                    cubes = 0;
                }
            }
            answer += (red * blue * green);
        }

        return answer;
    }

    private ArrayList<String> getGames() throws IOException {
        ArrayList<String> games = ReadLines("C:/Users/CameronGraybill/Documents/Code/AoC2023Java/Inputs/Day02Input.txt");
        return games;
    }
}

File: TypeScript/src/2022/Solvers/Day02Solver.ts
import { parseToStrings } from "../Helpers/InputParser.js";

var rounds = parseToStrings("TypeScript/src/2022/Inputs/Day02Input.txt");

function day02Part01() {
    var score = 0;

    var possibleOutcomes = {
        "A X": 4,
        "A Y": 8,
        "A Z": 3,
        "B X": 1,
        "B Y": 5,
        "B Z": 9,
        "C X": 7,
        "C Y": 2,
        "C Z": 6,
    }

    for (const round of rounds) {
        if (Object.keys(possibleOutcomes).includes(round)) {
            var points = possibleOutcomes[round];
            score += points;
        }
    }

    return score;
}

function day02Part02() {
    var score = 0;

    var possibleOutcomes = {
        "A X": 3,
        "A Y": 4,
        "A Z": 8,
        "B X": 1,
        "B Y": 5,
        "B Z": 9,
        "C X": 2,
        "C Y": 6,
        "C Z": 7,
    }

    for (const round of rounds) {
        if (Object.keys(possibleOutcomes).includes(round)) {
            var points = possibleOutcomes[round];
            score += points;
        }
    }

    return score;
}

console.log(day02Part02());